src/topics/async.md

```diff
@@ -0,0 +1,430 @@
+Asynchronous Programming
+========================
+How we perform more than one simultaneous action in Node.js.
+
+Audience: javascript programmers new to node.js.
+
+
+
+What Is Async Programming
+-------------------------
+Asynchronous programming is any technique in which the result of an operation
+is available outside of the normal synchronous (statement by statement) flow
+of code in a program.
+
+For example, you might call the function `foo()` and it returns somewhat
+quickly, but the result is available sometime _after_ the function has
+returned. The `foo()` function will provide a way for you to access this
+result.
+
+There are several techniques for asynchronous programming. Listed below are
+the ones commonly used in node.js. Sometimes more than one technique will be
+used within the same program.
+
+
+
+Why Async Is Used in Node.js
+----------------------------
+Node.js runs javascript on the server, and was created to be able to write
+long-running processes which handle multiple requests. These processes are
+sometimes called "apps" or "daemons", or even themselves called "servers".
+
+For example, a process which handles HTTP (web) requests might take some time
+to assemble each page out of various sources. If fetching data from these
+sources is done asynchronously then the process can listen for and start
+handling other requests while waiting for the sources to return their results.
+
+As well, asynchronous programming makes it easy to fetch data from multiple
+sources at the same time. This is faster than querying each source one at a
+time and waiting for each to complete before starting the next.
+
+By using asynchronous programming, well-written node.js processes can handle
+hundreds or thousands of requests per second. Fewer processes doing more work
+can be more efficient than many processes waiting idly for external sources to
+return.
+
+
+
+Async Approach: Callbacks
+-------------------------
+In this approach, when a function is called it is passed a function object as
+an argument. Once the result is available it is passed to this function
+object. The function object is often called a "callback" since it is "called"
+to return the result "back" to the caller.
+
+Consider the following example:
+
+```js
+function add(a, b, callback) {
+  setTimeout(function() {
+    var c = a + b;
+    callback(c);
+  }, 100);
+}
+
+add(12, 5, function(sum) {
+  console.log(sum);
+});
+```
+
+When we call `add()` we pass an anonymous function as the third argument. Once
+the sum is available the anonymous function will be called with it.
+
+There are two common refinements to the above technique: events and streams.
+
+
+### Events
```

Comment:
Hmmm... that's a good point. The "callback" idea is a little messy, since it's often used to refer to a function object which is called once (to return a value), but can also mean any function object which is called as appropriate (not necessarily "to return the final result of an operation").